packages/main/src/components/AnalyticalTable/pluginHooks/useIndeterminateRowSelection.tsx

```diff
@@ -39,46 +32,70 @@ export const useIndeterminateRowSelection = () => {
   };
 
   const stateReducer = (newState, action, prevState, instance) => {
+    const { rowsById, state, rows } = instance;
     if (action.type === 'INDETERMINATE_ROW_IDS') {
       if (action.payload === 'reset') {
         return {
           ...newState,
           indeterminateRows: {}
         };
       }
-      let indeterminateRows = {};
-      const allSelectedRows = {};
-      let allSelected = true;
-      let currentDepth = -1;
 
-      instance.flatRows
-        ?.filter((item) => !item.original.emptyRow)
-        .sort((a, b) => b.id.localeCompare(a.id))
-        .map((item) => {
-          if (currentDepth === -1) {
-            currentDepth = item.depth;
-          } else if (currentDepth !== item.depth) {
-            currentDepth = item.depth;
-            if (allSelected && newState.selectedRowIds[item.id]) {
-              allSelectedRows[item.id] = true;
-              delete indeterminateRows[item.id];
-            }
-            allSelected = true;
-          }
+      const indeterminateRowsById = {};
+      let usedParentIndex = '';
+      const getIndeterminate = (rows, rowIdScope = null) => {
+        rows.forEach((row) => {
+          if (row.subRows.length > 0) {
+            // find leaf nodes
+            getIndeterminate(row.subRows, row.id);
+          } else if (rowIdScope !== null && usedParentIndex !== rowIdScope) {
+            usedParentIndex = rowIdScope;
+            const checkIndeterminate = (rowId) => {
+              const getParentRow = (id) => {
+                let lastDotIndex = id.lastIndexOf('.');
+                if (lastDotIndex === -1) {
+                  lastDotIndex = Infinity;
+                }
+                const parentRowId = id.slice(0, lastDotIndex);
+                return [rowsById[parentRowId], lastDotIndex];
+              };
```

Comment:
we don't need to recreate this function every time - can be extracted to the `stateReducer` level